File: 2020/day03-2.ts
function solution(content: string) {
  const treemap = content.split("\n");
  return (
    countTrees(treemap, 0, 0, 1, 1) *
    countTrees(treemap, 0, 0, 3, 1) *
    countTrees(treemap, 0, 0, 5, 1) *
    countTrees(treemap, 0, 0, 7, 1) *
    countTrees(treemap, 0, 0, 1, 2)
  );
}

function countTrees(
  treemap: string[],
  startX: number,
  startY: number,
  deltaX: number,
  deltaY: number
) {
  const width = treemap.at(0)?.length ?? 0;
  const height = treemap.length;
  let currentX = startX;
  let currentY = startY;
  let trees = 0;
  while (currentY < height) {
    if (treemap[currentY][currentX] === "#") trees += 1;
    currentX = (currentX + deltaX) % width;
    currentY += deltaY;
  }
  return trees;
}

Deno.readTextFile("day03.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2020/day03-2.hs
import System.IO

solution :: String -> Int
solution content = product . map (\(dx, dy) -> countTrees (0, 0) (dx, dy) (lines content)) $ [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]

countTrees :: (Int, Int) -> (Int, Int) -> [String] -> Int
countTrees (startX, startY) (deltaX, deltaY) treemap
  | startY < length treemap = if' ((!! startX) ((!! startY) treemap) == '#') 1 0 + countTrees ((startX + deltaX) `mod` length ((!! 0) treemap), startY + deltaY) (deltaX, deltaY) treemap
  | otherwise = 0

if' :: Bool -> a -> a -> a
if' True x _ = x
if' False _ y = y

main :: IO ()
main = do openFile "day03.dat" ReadMode >>= hGetContents >>= print . solution
